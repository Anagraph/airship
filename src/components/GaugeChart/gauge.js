import React, { Component } from 'react';
import PropTypes from 'prop-types';
import GaugeChart from './gaugeChart';

class GaugeWidget extends Component {
  static defaultProps = {
    value: 50,
    maxValue: 100,
    minValue: 0,
    label: '',
    backgroundColor: '#F5F5F5',
    foregroundColor: '#47DB99',
    textColor: '#2C2C2C',
  };

  static propTypes = {
    value: PropTypes.number,
    maxValue: PropTypes.number,
    minValue: PropTypes.number,
    label: PropTypes.string,
    backgroundColor: PropTypes.string,
    foregroundColor: PropTypes.string,
    textColor: PropTypes.string,
  };

  componentDidMount() {
    this._chart = new GaugeChart(this.rootNode, this.props);
  }

  componentDidUpdate() {
    this._chart.update(this.props);
  }

  render() {
    return (
      <div ref={node => { this.rootNode = node; }} />
    );
  }
}

export default GaugeWidget;
