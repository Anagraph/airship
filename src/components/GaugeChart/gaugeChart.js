import { arc } from 'd3-shape';
import { select } from 'd3-selection';
import { interpolate } from 'd3-interpolate';
import { transition } from 'd3-transition'; // eslint-disable-line

const { PI } = Math;

const DEFAULT_OPTIONS = {
  width: 228,
  height: 100,
  innerRadius: 88,
  outerRadius: 100,
  maxValue: 180,
  minValue: 0,
  value: 0,
  backgroundColor: '#F5F5F5',
  foregroundColor: '#47DB99',
  textColor: '#2C2C2C',
};

export default class Gaugechart {
  constructor(element, options) {
    if (!element) throw new Error('A root container is required');

    this.element = element;
    this.options = Object.assign({}, DEFAULT_OPTIONS, options);

    this.createChart();
  }

  createChart() {
    this.arc = arc()
      .innerRadius(this.options.innerRadius)
      .outerRadius(this.options.outerRadius)
      .startAngle(-90 * (PI / 180))
      .cornerRadius(30);

    const svg = select(this.element)
      .append('svg')
      .attr('width', this.options.width)
      .attr('height', this.options.height)
      .append('g')
      .attr('transform', `translate(${this.options.width / 2},${this.options.height})`);

    this.background = svg
      .append('path')
      .datum({ endAngle: 90 * (PI / 180) })
      .style('fill', this.options.backgroundColor)
      .attr('d', this.arc);

    this.foreground = svg
      .append('path')
      .datum({ endAngle: -90 * (PI / 180) })
      .style('fill', this.options.foregroundColor)
      .attr('d', this.arc);

    this.valueLabel = svg
      .append('text')
      .attr('transform', 'translate(0, -30)')
      .attr('text-anchor', 'middle')
      .style('font-size', '40')
      .style('font-weight', '300')
      .style('color', this.options.textColor)
      .style('font-family', 'Roboto')
      .style('-webkit-font-smoothing', 'antialiased')
      .text(this.options.value);

    this.textLabel = svg
      .append('text')
      .attr('transform', 'translate(0, -5)')
      .attr('text-anchor', 'middle')
      .style('font-size', '12')
      .style('font-weight', '300')
      .style('color', this.options.textColor)
      .style('font-family', 'Roboto')
      .text(this.options.label);

    this._refresh();
  }

  update(options) {
    this.options = { ...this.options, ...options };

    this._refresh();
  }

  _refresh() {
    const value = (this.options.value * 180) / this.options.maxValue;
    const numPi = Math.floor(value - 90) * (PI / 180);

    this.textLabel.text(this.options.label);

    this.valueLabel
      .transition()
      .text(Math.floor(this.options.value));

    this.foreground.transition()
      .duration(750)
      .call(this._animateValue, numPi, this.arc);
  }

  _animateValue = (transition, newAngle, arcFn) => {
    transition.attrTween('d', d => {
      const interpolateFn = interpolate(d.endAngle, newAngle);

      return t => {
        d.endAngle = interpolateFn(t); // eslint-disable-line no-param-reassign

        return arcFn(d);
      };
    });
  }
}
