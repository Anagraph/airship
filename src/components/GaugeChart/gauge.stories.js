import React from 'react';
import { storiesOf } from '@storybook/react';
import { ThemeProvider } from 'styled-components';
import GaugeChart from './gauge';
import Widget from '../Widget/widget';
import { theme } from '../../constants';

const CUSTOM_THEME = {
  ...theme,
  type01: '#7E78E2',
};

class GaugeChartUpdated extends React.Component {
  state = {
    value: 60,
  }

  changeState = () => {
    this.setState({ value: parseInt(Math.random(100) * 100, 10) });
  }

  render() {
    return (
      <div>
        <GaugeChart value={this.state.value} />
        <button onClick={this.changeState}>Click me</button>
      </div>
    );
  }
}

storiesOf('GaugeChart', module)
  .add('Default', () => (
    <GaugeChart value={60} label="Test" />
  ))
  .add('Inside a widget', () => (
    <Widget>
      <Widget.Title>Title</Widget.Title>
      <Widget.Description>Description</Widget.Description>
      <GaugeChart value={60} />
    </Widget>
  ))
  .add('Updating data', () => (
    <Widget>
      <Widget.Title>Suffer score</Widget.Title>
      <Widget.Description>Just a widget</Widget.Description>

      <GaugeChartUpdated />
    </Widget>
  ))
  .add('With custom theme', () => (
    <ThemeProvider theme={CUSTOM_THEME}>
      <Widget>
        <Widget.Title>Suffer score</Widget.Title>
        <Widget.Description>Just a widget</Widget.Description>

        <GaugeChart value={12} label="parsecs" color="#7E78E2" />
      </Widget>
    </ThemeProvider>
  ));
