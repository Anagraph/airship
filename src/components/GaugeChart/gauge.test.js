import React from 'react';
import renderer from 'react-test-renderer';
import GaugeChart from './gauge';

describe('<GaugeChart />', () => {
  describe('render', () => {
    describe('when data passed in', () => {
      it('renders the correct amount of bars', () => {
        const component = renderer.create(<GaugeChart value={12} />);
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
      });

      it('renders with a label', () => {
        const component = renderer.create(<GaugeChart value={12} label="Parsecs" />);
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
      });

      it('renders with max and min values', () => {
        const component = renderer.create(
          <GaugeChart
            minValue={0}
            maxValue={2000}
            value={1337}
            label="Leet"
          />
        );
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
      });
    });
  });

  describe('update', () => {
    describe('when data changes', () => {
      it('should call the update method or the chart', () => {
        const component = renderer.create(<GaugeChart value={12} />);

        expect(component.toJSON()).toMatchSnapshot();

        // Changing properties should trigger a componentDidUpdate
        component.update(<GaugeChart value={70} />);

        expect(component.toJSON()).toMatchSnapshot();
      });
    });
  });
});
