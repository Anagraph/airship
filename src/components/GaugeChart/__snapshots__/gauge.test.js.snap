// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GaugeChart /> render when data passed in renders the correct amount of bars 1`] = `
.c0 {
  color: #2C2C2C;
}

.c1 {
  color: #2C2C2C;
  margin: 0;
  font-family: Roboto,sans-serif;
  font-weight: 400;
  -webkit-font-smoothing: antialiased;
  -webkit-transform: translate(50%,65%);
  -ms-transform: translate(50%,65%);
  transform: translate(50%,65%);
  font-size: 40px;
  font-weight: 300;
  text-anchor: middle;
  fill: #2C2C2C;
}

<svg
  className="c0"
  viewBox="0 0 228 100"
>
  <text
    className="c1"
  >
    12
  </text>
  
</svg>
`;

exports[`<GaugeChart /> render when data passed in renders with a label 1`] = `
.c0 {
  color: #2C2C2C;
}

.c2 {
  color: #2C2C2C;
  margin: 0;
  font-family: Roboto,sans-serif;
  font-weight: 400;
  -webkit-font-smoothing: antialiased;
  -webkit-transform: translate(50%,85%);
  -ms-transform: translate(50%,85%);
  transform: translate(50%,85%);
  font-size: 12px;
  font-weight: 300;
  text-anchor: middle;
  fill: #2C2C2C;
}

.c1 {
  color: #2C2C2C;
  margin: 0;
  font-family: Roboto,sans-serif;
  font-weight: 400;
  -webkit-font-smoothing: antialiased;
  -webkit-transform: translate(50%,65%);
  -ms-transform: translate(50%,65%);
  transform: translate(50%,65%);
  font-size: 40px;
  font-weight: 300;
  text-anchor: middle;
  fill: #2C2C2C;
}

<svg
  className="c0"
  viewBox="0 0 228 100"
>
  <text
    className="c1"
  >
    12
  </text>
  <text
    className="c2"
  >
    Parsecs
  </text>
</svg>
`;

exports[`<GaugeChart /> render when data passed in renders with max and min values 1`] = `
.c0 {
  color: #2C2C2C;
}

.c2 {
  color: #2C2C2C;
  margin: 0;
  font-family: Roboto,sans-serif;
  font-weight: 400;
  -webkit-font-smoothing: antialiased;
  -webkit-transform: translate(50%,85%);
  -ms-transform: translate(50%,85%);
  transform: translate(50%,85%);
  font-size: 12px;
  font-weight: 300;
  text-anchor: middle;
  fill: #2C2C2C;
}

.c1 {
  color: #2C2C2C;
  margin: 0;
  font-family: Roboto,sans-serif;
  font-weight: 400;
  -webkit-font-smoothing: antialiased;
  -webkit-transform: translate(50%,65%);
  -ms-transform: translate(50%,65%);
  transform: translate(50%,65%);
  font-size: 40px;
  font-weight: 300;
  text-anchor: middle;
  fill: #2C2C2C;
}

<svg
  className="c0"
  viewBox="0 0 228 100"
>
  <text
    className="c1"
  >
    1337
  </text>
  <text
    className="c2"
  >
    Leet
  </text>
</svg>
`;

exports[`<GaugeChart /> update when data changes should call the update method or the chart 1`] = `
.c0 {
  color: #2C2C2C;
}

.c1 {
  color: #2C2C2C;
  margin: 0;
  font-family: Roboto,sans-serif;
  font-weight: 400;
  -webkit-font-smoothing: antialiased;
  -webkit-transform: translate(50%,65%);
  -ms-transform: translate(50%,65%);
  transform: translate(50%,65%);
  font-size: 40px;
  font-weight: 300;
  text-anchor: middle;
  fill: #2C2C2C;
}

<svg
  className="c0"
  viewBox="0 0 228 100"
>
  <text
    className="c1"
  >
    12
  </text>
  
</svg>
`;

exports[`<GaugeChart /> update when data changes should call the update method or the chart 2`] = `
.c0 {
  color: #2C2C2C;
}

.c1 {
  color: #2C2C2C;
  margin: 0;
  font-family: Roboto,sans-serif;
  font-weight: 400;
  -webkit-font-smoothing: antialiased;
  -webkit-transform: translate(50%,65%);
  -ms-transform: translate(50%,65%);
  transform: translate(50%,65%);
  font-size: 40px;
  font-weight: 300;
  text-anchor: middle;
  fill: #2C2C2C;
}

<svg
  className="c0"
  viewBox="0 0 228 100"
>
  <text
    className="c1"
  >
    70
  </text>
  
</svg>
`;
